Exosuit Fabricator tgui total conversion - An Ode to Roboticists

Timberpoes opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Total conversion of the old HTML Exofab to a shiny, new tgui interface.

Includes a tiny bug fix here or there. Mining Scanner only worked if attached to "Working" mech models. It could be attached to other mechs but did nothing. It can now only be attached to "Working" mechs - Clarke and Ripley.

Touched up a few datum names and the odd description. Added a few bit flags to help categorise design datums. Added a special tag to design datums so they can have some additional search metadata (currently only used to allow people to search for the B.O.R.I.S. module by just using "boris").

![MvCfM9vMAu](https://user-images.githubusercontent.com/24975989/86781741-26dd9e00-c056-11ea-8671-1f7e59c46100.gif)

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Replaces the HTML monstrosity with a tgui monstrosity.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Roboticists rejoice. NanoTrasen has heard your cries for help. They've seen CCTV footage (recorded by the AI) of you all staring enviously at geneticists, lording around their shiny, polished, usable DNA Consoles. Cry no more, for you can now please your Silicon Overlords and robust those Machine Cultist Chaplains in high definition glory with a beautiful new interface added to Exosuit Fabricators.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Timberpoes committed:
First draft

Timberpoes committed:
tgui interface bluescreen guards

Timberpoes committed:
Preparing for cyborg module sub-categories

Timberpoes committed:
The rest of the owl

Timberpoes committed:
Stop duplicate say messgaes

Timberpoes committed:
Implement interface handling for obstructed exits

Timberpoes committed:
Pocket lint

Timberpoes committed:
Documentation and misc. code cleanup

Timberpoes committed:
Implement searching in interface

Timberpoes committed:
tgui pocket lint and build

Timberpoes committed:
Coding style consistency and changes. Rebuild tgui.

GunHog commented:
It is so beautiful...you deserve a gold star!

KathyRyals commented:
Literally a game changer for robotics. Timberpoes maintainer when ?

frontline03 commented:
Very Bay like, I like it.

dragomagol commented:
The addition of a search is a great QoL improvement. Looking forward to playing roboticist with this!

silicons commented:
based

Kaffe-work commented:
oh my god, no more old shit exofabs? glorious

ATH1909 commented:
holy shit YES

TheObserver-sys commented:
This right here? This is the good shit. Fantastic work.

Spepbuthewranglesgithub commented:
by the omnissiah its wonderful

Flakkyboo commented:
Will the production queue display total material cost? If it doesn't I would really like it to so ill know if my big queue of parts can be made all in one

zombiegums commented:
Absolutely based.

GunHog commented:
> Will the production queue display total material cost? If it doesn't I would really like it to so ill know if my big queue of parts can be made all in one

It will! The final part of the demo shows the total materials needed for the items in the production queue, and will show red if you do not have enough. It really is all you could ever want in an Exofab UI!

Fikou commented on code/game/mecha/equipment/mecha_equipment.dm:
```diff
@@ -13,6 +13,8 @@
 	///Bitflag. Determines the range of the equipment.
 	var/range = MECHA_MELEE
+	/// Bitflag. Used by exosuit fabricator to assign sub-categories based on which exosuits can equip this.
+	var/mech_flags = 0
```
NONE?

Timberpoes replied:
Done

Fikou commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -11,6 +11,8 @@
 	var/require_module = 0
 	var/list/module_type = null
+	///	Bitflags listing module compatibility. Used in the exosuit fabricator for creating sub-categories.
+	var/list/module_flags = 0
```
NONE?

Timberpoes replied:
Done

Timberpoes committed:
Update default bitflags

Timberpoes committed:
Address potential edge case bluescreen.

Timberpoes committed:
orange man good

actioninja commented:
dnm until tgui 4.0 is resolved. More to follow.

actioninja reviewed (dismissed):
Looks more or less fine with some minor quibbles and concerns

actioninja commented:
Since tgui 4.0 is coming up very soon and/or getting shelved again, I'm tentatively throwing a dnm on this until that's handled.

If it does get merged, here's the needed refactors:

- Call signature change in ui_interact
- Different handling of assets (new `ui_assets`)
- Move ui x and y to frontend

Timberpoes commented:
Understood. I'll update to tgui 4.0 compatbility in the coming days.

Timberpoes commented on code/game/mecha/mech_fabricator.dm:
```diff
@@ -62,11 +88,15 @@
 		. += "<span class='notice'>The status display reads: Storing up to <b>[rmat.local_size]</b> material units.<br>Material consumption at <b>[component_coeff*100]%</b>.<br>Build time reduced by <b>[100-time_coeff*100]%</b>.</span>"
 
+
+// TODO - Delete me, I no longer work.
```
Looks like I forgot my TODO sweep.

Removing the required access doesn't do anything (to my knowledge, I don't believe the current Exofab does anything when emagged either) - Rather than implement it, I had it slated for removal pending future ideas for more interesting emag interations. Turns out I never deleted it, duh-doy.

Timberpoes commented on tgui/packages/tgui/interfaces/ExosuitFabricator.js:
```diff
@@ -0,0 +1,780 @@
+};
+
+const queueCondFormat = (materials, queue) => {
+  let material_tally = {};
+  let mat_format = {};
+  let missing_mat_tally = {};
+  let text_colors = {};
```
I can run through a simple regex and replace to camelCase the names up easily enough. I'll get that done in the tgui 4.0 compatability commit.

coolsurf6 commented:
This
Is
Beautiful

PKPenguin321 commented:
I feel like the queue doesn't need to be so wide, it's the same width as the entire parts selection area when it doesn't need as much space as it. You could thin it down a little bit and give the part selection area a bit more breathing room.

Timberpoes commented:
> 
> 
> I feel like the queue doesn't need to be so wide, it's the same width as the entire parts selection area when it doesn't need as much space as it. You could thin it down a little bit and give the part selection area a bit more breathing room.

Queue length is based on the (current) worst-case scenario crafting recipe name not line wrapping.

I didn't show in the image, however, the middle section part listing is flexible in size and will grow or shrink to accomodate the window size.

Timberpoes commented:
With tgui 4.0 merged I'm not going to work on fixing all the various merge conflicts created in the interim and bring this back up to tgui 4.0 standards

Timberpoes committed:
Remove emag_act with obsolete functionality

Timberpoes committed:
Standardise variable naming conventions

Timberpoes committed:
tgui 4.0 update

Timberpoes committed:
tgui 4.0 update

Timberpoes committed:
Port a couple of usability improvements from concept branch.

Timberpoes committed:
Fix new ui_act input.

Timberpoes committed:
Actually fix new ui_act input

Timberpoes committed:
Port some small improvements from concept branch.

Timberpoes committed:
Rebuild tgui

stylemistake commented on code/game/mecha/mech_fabricator.dm:
```diff
@@ -458,23 +452,23 @@
+		ui = new(user, src, "ExosuitFabricator")
 		ui.open()
 
+/obj/machinery/mecha_part_fabricator/ui_state()
+	return GLOB.default_state
+
```
Redundant proc, because `GLOB.default_state` is well.... default.

Timberpoes committed:
Made department of redundancy department redundant.

actioninja commented:
4.0 is merged now, clearing dnm

Timberpoes committed:
Rebuild tgui.

Timberpoes commented:
Ported to tgui 4.0.

@actioninja 
Previous code review has been addressed. I've removed the emag_act proc as it doesn't do anything anymore (unless I'm mistaken, it doesn't do anything on live either?) JS variable naming is now more consistently camelCase.

@stylemistake 
Removed redundant ui_state proc
![image](https://user-images.githubusercontent.com/24975989/87732098-f0292580-c7c3-11ea-808d-0da0dab773b8.png)
